cheatsheets/Microservices_Security_Cheat_Sheet.md

```diff
@@ -2,176 +2,660 @@
+
+Although this pattern improves observability and supports a broader range of access control models, it introduces its own trade-offs and does not eliminate all challenges found in the previous pattern.
+
+#### Pros
+
+* **Policy governance:** Policies can be centrally defined, versioned, reviewed, and audited, independent of the service’s implementation language.
+* **Policy layering:** The model allows for both global (e.g. security team–defined) and local (e.g. service team–defined) policies to coexist. This enables clearer separation of concerns and better alignment with organizational structure and responsibilities.
+* **Improved monitoring:** All decisions can be consistently logged and monitored, assuming proper instrumentation.
+* **Team autonomy:** Teams remain responsible for their services and their integration with the PDP. This aligns well with independent team ownership and domain-driven design principles.
+* **Support for "before-the-fact" audit:** Particularly with ReBAC and NGAC systems as PDP, authorization models allow querying the existing access rights making answering the corresponding questions a simple game.
+* **Support for additional PDP models:** This pattern enables the use of a broader set of access control models, including ReBAC and NGAC systems, which may better fit for a given business context.
+
+#### Cons
+
+* **Policy distribution complexity:** Policies are managed independently of service code, and the PDP may support multiple services and domains. Ensuring the correct policy version is applied consistently can be complex.
+* **Context sharing:** PDPs do not inherently have access to service context. Mechanisms must be designed to assemble and pass the right attributes into the PDP for evaluation and in case of the ReBAC/NGAC systems to build the actual authorization model.
+* **Performance overhead:** Network hops between the microservice and the external PDP introduce latency and a potential point of contention under high load.
+* **Incomplete enforcement observability:** While policy decisions are consistently logged by the PDP, there’s still limited visibility into whether those decisions were enforced correctly in the service code. Instrumentation gaps can hinder detection of abuse, validation of protections, and system-wide access analysis.
+* **Coverage gaps:** There is no guarantee that every service or endpoint consistently integrates with the central PDP. Some frameworks expose endpoints by default, often without offering hooks for policy enforcement. Teams may also just forget to add the required logic to some endpoints. Combined with poor or misleading documentation, this can result in unintentionally exposed functionality and missed access control.
```

Comment:
> Some frameworks expose endpoints by default, often without offering hooks for policy enforcement.
I stumble about this sentence now the second time, and it's not 100% clear for me what you mean here. Can you explain?

Reply:
E.g. Spring actuator endpoints, fastapi (python) exposed endpoints for the same purposes, or the openapi related endpoints. Static routes e.g. in express.js, etc